### PR TITLE
Rework help command to avoid a deepcopy on invoke

### PR DESCRIPTION
**Pull request by bijij**
## Summary

This PR is my first go-around at reworking help command to avoid a deepcopy.

I'm not 100% sure on the implementation here, but it's an option we can consider.

## Checklist

- [x] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/ext/commands/help.py [hunk 1]
```diff
@@ -498,31 +398,54 @@ def replace(obj: re.Match, *, transforms: Dict[str, str] = self.MENTION_TRANSFOR
...
+        self._cog = None
 
-        Returns
-        --------
-        Optional[:class:`Cog`]
-            The cog that is currently set for the help command.
-        """
-        return self._command_impl.cog
+    @property
+    def cog(self) -> Optional[Cog]:
```
Comment (Gobot1234) [inline discord/ext/commands/help.py:438]: Shouldn't this be CogT?

File: discord/ext/commands/help.py [hunk 2]
```diff
@@ -219,92 +219,12 @@ def _not_overridden(f: FuncT) -> FuncT:
...
-        # revert back into their original methods
-        cog = self.cog
-        cog.get_commands = cog.get_commands.__wrapped__
-        cog.walk_commands = cog.walk_commands.__wrapped__
-        self.cog = None
+_context: ContextVar[Optional[Context]] = ContextVar('context', default=None)
 
 
-class HelpCommand:
+class HelpCommand(Command[Optional['Cog'], ... if TYPE_CHECKING else Any, Any], Generic[ContextT]):
```
Comment (Rapptz) [inline discord/ext/commands/help.py:225]: Need to be in a separate line.

File: discord/ext/commands/help.py [hunk 3]
```diff
@@ -336,88 +256,57 @@ class HelpCommand:
...
-        __original_args__: Tuple[Any, ...]
-
     def __new__(cls, *args: Any, **kwargs: Any) -> Self:
-        # To prevent race conditions of a single instance while also allowing
-        # for settings to be passed the original arguments passed must be assigned
-        # to allow for easier copies (which will be made when the help command is actually called)
-        # see issue 2123
+        # we need to wrap the command callback to ensure context is correctly set.
+        functools.update_wrapper(cls._set_context, cls.command_callback)
+        cls._set_context.__qualname__ = 'command_callback'
```
Comment (Rapptz) [inline discord/ext/commands/help.py:262]: Not really a fan of this. I know this is attempting to lie to `utils.is_inside_class`. I have no idea what the wrapping is for.

File: discord/ext/commands/context.py [hunk 4]
```diff
@@ -354,52 +354,47 @@ async def send_help(self, *args: Any) -> Any:
...
             return None
 
-        cmd = cmd.copy()
-        cmd.context = self  # type: ignore
+        _context.set(self)
+
         if len(args) == 0:
             await cmd.prepare_help_command(self, None)
-            mapping = cmd.get_bot_mapping()
+            entity = cmd.get_bot_mapping()
```
Comment (Rapptz) [inline discord/ext/commands/context.py:369]: This change doesn't seem right.

**Review by Rapptz — CHANGES_REQUESTED** [review]

File: discord/ext/commands/help.py (hunk 3 above)
Comment (bijij) [reply discord/ext/commands/help.py:262]: Wrapping is to have `self.context` set without requring subclasses to explicitly do so.

This was one of my own main gripes with the implimentation.

File: discord/ext/commands/context.py (hunk 4 above)
Comment (bijij) [reply discord/ext/commands/context.py:369]: `HelpCommand.send_bot_help` excepts a `Mapping`, so it's correct. 
In an earlier version of this PR i needed to make changes to how the call was done which I no longer do, so technically this refactor of `send_help` could be undone.

File: discord/ext/commands/help.py (hunk 3 above)
Comment (bijij) [reply discord/ext/commands/help.py]: I reverted back to overriding params in the constructor.

**Review by Rapptz — APPROVED** [review]

**Pull request merged**